File: P11 Container With Most Water/Solution.java
class Solution {
    public int maxArea(int[] height) {
        return maxArea2(height);
    }
    
    // Approach 1: Brute Force
    // Time: O(N ^ 2)
    // Space: O(1)
    private int maxArea1(int[] height) {
        int result = 0;
        // left pointer
        for (int i = 0; i < height.length; i++) {
            // right pointer
            for (int j = i + 1; j < height.length; j++) {
                result = Math.max(result, Math.min(height[i], height[j]) * (j - i));
            }
        }
        return result;
    }
    
        
    // Approach 2: Two pointers
    // Time: O(N)
    // Space: O(1)
    private int maxArea2(int[] height) {
        int result = 0;
        int left = 0, right = height.length - 1;
        while (left != right) {
            result = Math.max(result, Math.min(height[left], height[right]) * (right - left));
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return result;
    }
    
}

File: P11 Container With Most Water/solution.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        return self._maxArea2(height)
    
    # Approach 1: Brute Force
    # Time: O(N ^ 2)
    # Space: O(1)
    def _maxArea1(self, height: List[int]) -> int:
        result = 0
        for i in range(0, len(height)):
            for j in range(i + 1, len(height)):
                result = max(result, min(height[i], height[j]) * (j - i))
        return result
    
    # Approach 2: Two pointers
    # Time: O(N)
    # Space: O(1)
    def _maxArea2(self, height: List[int]) -> int:
        result = 0
        left = 0
        right = len(height) - 1
        while left != right:
            result = max(result, min(height[left], height[right]) * (right - left))
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        return result 